web_poet/testing/pytest.py

```diff
@@ -34,38 +46,106 @@ def collect(self) -> Iterable[Union[pytest.Item, pytest.Collector]]:  # noqa: D1
-            nodeid: Optional[str] = None,
-        ) -> None:
-            super().__init__(name, parent, config, session, nodeid)
-            self.fixture = Fixture(Path(self.fspath, self.name))
-
-    def runtest(self) -> None:  # noqa: D102
-        self.fixture.assert_output()
+    def __init__(self, name: str, parent=None, **kwargs) -> None:
+        super(WebPoetCollector, self).__init__(name, parent, **kwargs)
+        self.fixture = Fixture(self._path)
+
+    def collect(self) -> Iterable[Union[pytest.Item, pytest.Collector]]:
+        """Return a list of children (items and collectors) for this
+        collection node."""
+        if self.config.getoption("WEB_POET_TEST_PER_ITEM", default=False):
+            return [
+                WebPoetItem.from_parent(parent=self, name="item", fixture=self.fixture)
+            ]
+        else:
+            field_tests = [
+                WebPoetFieldItem.from_parent(
+                    parent=self, name=field, fixture=self.fixture, field_name=field
+                )
+                for field in self.fixture.get_expected_output_fields()
+            ]
+            no_extra_fields_tests = [
+                WebPoetNoExtraFieldsItem.from_parent(
+                    parent=self, name="NO_EXTRA_FIELDS", fixture=self.fixture
+                )
+            ]
+            return field_tests + no_extra_fields_tests
+
+
+class _WebPoetItem(pytest.Item, _PathCompatMixin):
+    def __init__(self, *, fixture: Fixture, **kwargs) -> None:
+        super().__init__(**kwargs)
+        self.fixture = fixture
+
+
+class WebPoetItem(_WebPoetItem):
+    def runtest(self) -> None:
+        self.fixture.assert_full_item_correct()
+
+    def reportinfo(self):
+        return self._path, 0, f"{self.fixture.short_name}"
+
+    def repr_failure(self, excinfo, style=None):
+        if isinstance(excinfo.value, ItemValueError):
+            got, expected = excinfo.value.args
+            return comparison_error_message(
+                config=self.config,
+                op="==",
+                expected=expected,
+                got=got,
+                prefix="The output doesn't match.",
+            )
+        else:
+            return super().repr_failure(excinfo, style)
+
+
+class WebPoetNoExtraFieldsItem(_WebPoetItem):
+    def runtest(self) -> None:
+        self.fixture.assert_no_extra_fields()
+
+    def reportinfo(self):
+        return self._path, 0, f"{self.fixture.short_name}: extra fields"
+
+    def repr_failure(self, excinfo, style=None):
+        if isinstance(excinfo.value, FieldsUnexpected):
+            fields = excinfo.value.args[0]
+            return f"The item contains unexpected fields: \n{self._format_extra_fields(fields)}"
+        else:
+            return super().repr_failure(excinfo, style)
+
+    def _format_extra_fields(self, extra_fields):
+        lines = []
+        for field, value in extra_fields.items():
+            lines.append(f" * {field} == {value!r}")
```

Comment:
```suggestion
            lines.append(f" * {field} = {value!r}")
```

Reply:
Should it be without whitespace around `=` then?